File: pkg/sdk/go/processing/generator.go
package processing

import (
	"encoding/json"
	"fmt"
	"os"
	"runtime"

	workflowv1 "github.com/kination/nautikus/api/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

// GenerateManifest creates the DAG JSON manifest from task definitions
func GenerateManifest(dagName string, tasks []TaskDef) {
	scriptContent := readCallerSource()

	taskSpecs := make([]workflowv1.TaskSpec, 0, len(tasks))

	for _, task := range tasks {
		spec := workflowv1.TaskSpec{
			Name:         task.Name,
			Type:         workflowv1.TaskTypeGo,
			Script:       scriptContent,
			Dependencies: task.Dependencies,
			Env: map[string]string{
				"NAUTIKUS_TASK_NAME": task.Name,
			},
		}

		// Add branch metadata for conditional tasks
		if task.BranchCondition != "" {
			spec.Env["NAUTIKUS_BRANCH_CONDITION"] = task.BranchCondition
			spec.Env["NAUTIKUS_CONDITION_SOURCE"] = task.ConditionSource
		}

		// Mark branch selector tasks
		if task.TaskType == TaskTypeBranch {
			spec.Env["NAUTIKUS_TASK_TYPE"] = "branch"
			// Store branch targets for controller to use
			if len(task.BranchTargets) > 0 {
				spec.Env["NAUTIKUS_BRANCH_TARGETS"] = joinStrings(task.BranchTargets, ",")
			}
		}

		// Mark join tasks
		if task.TaskType == TaskTypeJoin {
			spec.Env["NAUTIKUS_TASK_TYPE"] = "join"
		}

		taskSpecs = append(taskSpecs, spec)
	}

	dag := workflowv1.Dag{
		TypeMeta: metav1.TypeMeta{
			APIVersion: "workflow.nautikus.io/v1",
			Kind:       "Dag",
		},
		ObjectMeta: metav1.ObjectMeta{
			Name: dagName,
		},
		Spec: workflowv1.DagSpec{
			Tasks: taskSpecs,
		},
	}

	output, err := json.MarshalIndent(dag, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error marshaling DAG: %v\n", err)
		os.Exit(1)
	}

	fmt.Println(string(output))
}

func readCallerSource() string {
	// Walk up the stack to find the original caller (user's dag file)
	for i := 1; i < 10; i++ {
		_, callerFile, _, ok := runtime.Caller(i)
		if !ok {
			break
		}
		// Skip internal SDK files
		if !isSDKFile(callerFile) {
			content, err := os.ReadFile(callerFile)
			if err == nil {
				return string(content)
			}
		}
	}

	// Fallback: try common paths
	wd, _ := os.Getwd()
	paths := []string{
		wd + "/test/dags/go_dag.go",
	}
	for _, path := range paths {
		content, err := os.ReadFile(path)
		if err == nil {
			return string(content)
		}
	}

	fmt.Fprintf(os.Stderr, "Error: could not read source file\n")
	os.Exit(1)
	return ""
}

func isSDKFile(path string) bool {
	return len(path) > 0 && (contains(path, "/sdk/go/") || contains(path, "/sdk/python/"))
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(s) > 0 && containsHelper(s, substr))
}

func containsHelper(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

func joinStrings(strs []string, sep string) string {
	if len(strs) == 0 {
		return ""
	}
	result := strs[0]
	for i := 1; i < len(strs); i++ {
		result += sep + strs[i]
	}
	return result
}

File: pkg/sdk/python/processing/generator.py
"""DAG manifest generation logic"""
import json
import sys
import inspect
from typing import List

from .types import TaskDef, TaskType


def generate_manifest(dag_name: str, tasks: List[TaskDef]):
    """Create the DAG JSON manifest from task definitions"""
    script_content = _read_caller_source()

    task_specs = []

    for task in tasks:
        spec = {
            "name": task.name,
            "type": "Python",
            "script": script_content,
            "dependencies": task.dependencies,
            "env": {
                "NAUTIKUS_TASK_NAME": task.name
            }
        }

        # Add branch metadata for conditional tasks
        if task.branch_condition:
            spec["env"]["NAUTIKUS_BRANCH_CONDITION"] = task.branch_condition
            spec["env"]["NAUTIKUS_CONDITION_SOURCE"] = task.condition_source

        # Mark branch selector tasks
        if task.task_type == TaskType.BRANCH:
            spec["env"]["NAUTIKUS_TASK_TYPE"] = "branch"
            if task.branch_targets:
                spec["env"]["NAUTIKUS_BRANCH_TARGETS"] = ",".join(task.branch_targets)

        # Mark join tasks
        if task.task_type == TaskType.JOIN:
            spec["env"]["NAUTIKUS_TASK_TYPE"] = "join"

        task_specs.append(spec)

    manifest = {
        "apiVersion": "workflow.nautikus.io/v1",
        "kind": "Dag",
        "metadata": {
            "name": dag_name
        },
        "spec": {
            "tasks": task_specs
        }
    }

    print(json.dumps(manifest, indent=2))


def _read_caller_source() -> str:
    """Read the source file that called the SDK"""
    # Walk up the stack to find the user's dag file (not SDK files)
    for frame_info in inspect.stack():
        caller_file = frame_info.filename
        # Skip SDK files and built-in locations
        if not _is_sdk_file(caller_file) and not caller_file.startswith('<'):
            try:
                with open(caller_file, 'r') as f:
                    return f.read()
            except Exception:
                continue

    sys.stderr.write("Error: could not read source file\n")
    sys.exit(1)


def _is_sdk_file(path: str) -> bool:
    """Check if path is part of the SDK"""
    return '/sdk/python/' in path or '/sdk/go/' in path
